Add ISO3 categorization containing countries etc.

mikapfl opened the pull request:
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Changelog snippet in a `.rst` file in the directory `changelog_unreleased` added – remember to start with a `*` to make it a bullet point

## Description

Add a categorization containing countries and other areas from ISO 3166, country groups from UNFCCC and EU documentation, and their relationships.

Highlights:
* Parties to the UN Framework Convention on Climate Change, including Annex-I and Non-Annex-I parties.
* Uses data from the iso-codes package directly, so is more up-to-date than pycountry (as long as we update regularly).
* membership in the EU over time.

## Open questions
* Are the codes not in use in ISO 3166 fine? I added some stuff myself, in particular `Earth` and the groupings.
* How do we deal with the conventional `EU-{number_of_members}` codes? People out there use them, but `EU-27` is already unclear, and if someone else joins before the UK rejoins, `EU-28` will also be unclear. I used `EU_{year}` as primary codes instead, which is more robust, but included the (for now) unequivocal `EU-{number_of_members}` codes as well. However, they might break in future.
* Do we need other country groupings?
* Is ISO3 as name fine? We have used it in primap-hist for the last few releases, but it is honestly a stupid name. ISO 3 is something else, so ISO3166 would probably be a good name. However, if you google ISO 3, you find ISO3166 because nobody cares about some standard about standard numbering. What is more pain, a somewhat stupid name forever, or changing the name for future primap-hist releases?

## Context
We need this now in NDC Quantifications mostly for the Annex-I vs. Non-Annex-I info, but will probably benefit also more broadly.

lewisjared commented on climate_categories/tests/test_iso3.py:
```diff
@@ -0,0 +1,32 @@
+    assert len(climate_categories.ISO3["Annex-I"].children[0]) == 43
+    assert len(climate_categories.ISO3["Non-Annex-I"].children[0]) == 155
+    # leaf children excludes EU, because EU is not a leaf
+    assert len(climate_categories.ISO3["UNFCCC"].leaf_children[0]) == 197
```
In the NDC tool, we use 196 countries. What is the extra one here?

mikapfl replied:
The country which is a party to the UNFCCC but not in the NDC tool is :drum: :
```python
('VAT',
 {'title': 'Holy See (Vatican City State)',
  'alternative_codes': ['VA', '336']})
 ```

mikapfl replied:
They've been a party to the UNFCCC [since September last year already](https://unfccc.int/node/61239), try to keep up :wink:

JGuetschow replied:
The IAM groups like R5, RC10 are useful but have to be defined on a per IAM basis. I spend quite some time gathering this info for the SSP downscaling, but probably it has changed now as IAMs probably are now more detailed. So I think we should add them only when used.

An regarding changing the country terminology name from ISO3 to whatever. I prefer if it stays the same as that's probably hardcoded in a lot of places as I almost never use other country terminologies. But of course it's doable to change it.

lewisjared reviewed (approved):
> Are the codes not in use in ISO 3166 fine? I added some stuff myself, in particular Earth and the groupings.

I think that would be extremely useful given that is where climate_categories excels.

> How do we deal with the conventional EU-{number_of_members} codes? People out there use them, but EU-27 is already unclear, and if someone else joins before the UK rejoins, EU-28 will also be unclear. I used EU_{year} as primary codes instead, which is more robust, but included the (for now) unequivocal EU-{number_of_members} codes as well. However, they might break in future. 

Nice solution.

> Do we need other country groupings?

There are other groupings in `src/ndcs/data/country_info.csv` that might be useful although I'm not sure how authoritative they are. Not all of those groupings are useful so I don't think we need to do them all. Perhaps G7/G8 + G20. I think they are fairly consistent except for Russia joining and then being expelled from G7.

For example, the R5 regions would be great, but I'm not sure if there is a complete "authoritative" source other than https://tntcat.iiasa.ac.at/SspDb/dsd?Action=htmlpage&page=10. I know each IAM group may have different mapping to these categories, but I assume there is agreement on the country membership for each.

> Is ISO3 as name fine? We have used it in primap-hist for the last few releases, but it is honestly a stupid name. ISO 3 is something else, so ISO3166 would probably be a good name. However, if you google ISO 3, you find ISO3166 because nobody cares about some standard about standard numbering. What is more pain, a somewhat stupid name forever, or changing the name for future primap-hist releases?

It is used pretty widely as synonym for ISO 3166-1 so it is as good as any. It's a little confusing given that the canonical names are the alpha-3 dataset from ISO 3166-1 which is part of ISO 3166. Lots of 3's.

I don't know about the change management required for PRIMAP-hist to do a change in the region naming scheme. Is there a concept of alternative codes for datasets?

JGuetschow commented:
Very nice. I hope that I find time to replace my use of pycountry with this. What would be really helpful in the UNFCCC context would be a filed for the UNFCCC name which is not always the official name. I have not checked yet if there are cases where it's neither the name nor the official name given in the current specification. Also having old names e.g. for Türkiye would be great, so e.g. data which uses old names can be automatically mapped to the right country. It's all not urgent as I currently use workarounds for this as pycountry is not completely consistent with the UNFCCC country names and thus I need manual mapping for some countries anyway.

mikapfl commented:
I have added common names from the source and UNFCCC names where they were neither the short nor the official name. I took UNFCCC names from https://unfccc.int/process/parties-non-party-stakeholders/parties-convention-and-observer-states , I hope that covers them?

mikapfl commented:
Okay, I added G7, G8, G20, and OECD. Also added Turkey as a historical name for Türkiye, we can add more country groups and historical names when needed.
